src/esn.py

```diff
@@ -132,19 +179,181 @@ def _forward(self, ut: Array, x_init: Optional[Array] = None,
+            else:
+                y = np.dot(self.w_out, x)
+            yt.append(y)     
+            if collect_states:
+                xt.append(x)
+                zt.append(z)
+        # collect outputs and reservoir states into matrices
+        yt = np.concatenate(yt, axis=1).T
+        
+        if collect_states:
+            xt = np.concatenate(xt, axis=1).T
+            zt = np.concatenate(zt, axis=1).T
+            return xt, zt, yt
+        else:
+            return yt    
+        
+    def full_procedure(self, ut, trainingConfig: TrainingConfig):  
+        """
+        Run the whole procedure of loading and computing the conceptors.
+        
+        :param ut: (T, K)
+        :return Ci: conceptor matrix full (N,N) or scalar (N_rfc, 1)
+
+        """
+        
+        if self.random_feature:
+            xt, zt, yt = list(zip(*map(self.harvest_states, ut)))
+            # concatenate patterns along time
+            X = helper.concatenate_patterns(xt, trainingConfig.washout)
+            U = helper.concatenate_patterns(ut, trainingConfig.washout)
+            Y_T = U.copy()
+            self.update_weights(X, U, Y_T, alpha=trainingConfig.wout_regularizer)
+            xt_conceptor, zt_conceptor, yt_trained = list(zip(*map(self.harvest_states, ut)))
+            
+            Z_ = helper.concatenate_patterns(zt, trainingConfig.washout, shift=-1)
+            Z = helper.concatenate_patterns(zt, trainingConfig.washout, shift=0)
+            
+            D_loaded = ridge_regression(Z_, np.dot(self.w_in, U.T).T, regularizer=trainingConfig.d_regularizer)
+            G_loaded = ridge_regression(Z, np.dot(self.g, Z.T).T, regularizer=trainingConfig.g_regularizer)
+            self.g = G_loaded + D_loaded.copy()
+            
+            Ci = [
+                np.diag(compute_conceptor_diag(zt_i[trainingConfig.washout:, :], aperture=trainingConfig.aperture)).reshape(-1, 1)
+                for zt_i in zt
+            ]
+            
+        else:
+    
+            xt, yt = list(zip(*map(self.harvest_states, ut)))
+            # concatenate patterns along time
+            X = helper.concatenate_patterns(xt, trainingConfig.washout)
+            U = helper.concatenate_patterns(ut, trainingConfig.washout)
+        
+            Y_T = U.copy()
+            self.update_weights(X, U, Y_T, alpha=trainingConfig.wout_regularizer)
+        
+            X_ = helper.concatenate_patterns(xt, trainingConfig.washout, shift=-1)
+            B = np.repeat(self.b, X_.shape[0], axis=1).T
+        
+            W_loaded = loading_ridge_report(X, X_, B, regularizer=trainingConfig.w_regularizer)
+            
+            self.w = W_loaded.copy()
+        
+            Ci = [
+            compute_conceptor(xt_i[trainingConfig.washout:, :], aperture=trainingConfig.aperture)
+            for xt_i in xt
+            ]
+        return Ci
+    
+    def full_procedure_diag(self, ut, trainingConfig: TrainingConfig):  
```

Comment:
again, I would either
- merge this into the full_procedure function and do if statements for the diagonal conceptor specific parts
- call a function full_procedure which takes a parameter which decides which to call, e.g. full_procedure_rf or full_procedure_diag

Reply:
It seems misleading because the procedure of the diagonal is very different from the other one (it's Jong's ugly trick). Maybe just changing the name? After I plan to create a function specific to the different "full_procedure" (but they are more iterative than the classical one) for diagonal conceptors.